providers/implementations/encode_decode/encode_key2blob.c

```diff
@@ -73,10 +71,37 @@ static int key2blob_get_params(OSSL_PARAM params[])
     return 1;
 }
 
-static int key2blob_does_selection(void *vctx, int selection)
+static int key2blob_check_selection(int selection, int selection_mask)
 {
-    return (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0
-        && (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) == 0;
+    /*
+     * The selections are kinda sorta "levels", i.e. each selection given
+     * here is assumed to include those following.
+     */
+    int checks[] = {
+        OSSL_KEYMGMT_SELECT_PRIVATE_KEY,
+        OSSL_KEYMGMT_SELECT_PUBLIC_KEY,
+        OSSL_KEYMGMT_SELECT_ALL_PARAMETERS
+    };
+    size_t i;
+
+    /* The decoder implementations made here support guessing */
+    if (selection == 0)
+        return 1;
+
+    for (i = 0; i < OSSL_NELEM(checks); i++) {
+        int check1 = (selection & checks[i]) != 0;
+        int check2 = (selection_mask & checks[i]) != 0;
+
+        /*
+         * If the caller asked for the currently checked bit(s), return
+         * whether the decoder description says it's supported.
+         */
+        if (check1)
+            return check2;
```

Comment:
(FWIW, this is the behavior that I saw in other implementations and thought didn't quite match the documentation.  Given the note about "levels" I could sort-of see it making sense if we ran the checks in the other order -- if someone asks for pubkey or parameters, and we can't do parameters, we're not going to be able to do pubkey either.
OTOH, with `check1` and `check2` only asking if *any* bit is set (vs. exact match), the semantics are still surprising to me...)

Reply:
It sounds like I need to expand on the documentation for selectors.

Reply:
No need to prioritize it just on my account.
I'm just looking at this from the sense of "why are we making three passes?"  My understanding is that `OSSL_KEYMGMT_SELECT_PRIVATE_KEY` is just `OSSL_KEYMGMT_SELECT_ALL_PARAMETERS | <more bits>`, so that if the third pass through the loop was going to set `check1`, so is the first pass through the loop, so we would return before we ever got past the first pass.

Reply:
I wasn't planning on amending this PR, I was thinking of having another PR specifically for documentation.